server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileTransactionLog.java

```diff
@@ -0,0 +1,385 @@
+          }
+          this.updateCount = this.tabletLog.getUpdateCount();
+        }
+      }
+    }
+
+    public void clear() {
+      this.tabletLog.clear();
+      this.initialTs = System.currentTimeMillis();
+      this.initialFiles = this.finalFiles;
+      this.updateCount = this.tabletLog.getUpdateCount();
+    }
+
+    public void reset(Set<StoredTabletFile> files) {
+      this.tabletLog.clear();
+      this.initialTs = System.currentTimeMillis();
+      this.initialFiles = this.finalFiles = files.toArray(new StoredTabletFile[0]);
+      this.updateCount = this.tabletLog.getUpdateCount();
+    }
+
+    /**
+     * Add the passed in transaction, adjusting the log size as needed.
+     *
+     * @param transaction The new transaction
+     */
+    public void addTransaction(DatafileTransaction transaction) {
+      DatafileTransaction removed = this.tabletLog.add(transaction);
+      if (removed != null) {
+        this.initialTs = removed.ts;
+        this.initialFiles = applyTransaction(this.initialFiles, removed);
+      }
+      this.finalFiles = applyTransaction(this.finalFiles, transaction);
+      this.updateCount = this.tabletLog.getUpdateCount();
+    }
+
+    /**
+     * Apply a transaction to the set of files and return an updated file set
+     *
+     * @param files The initial files
+     * @param transaction The transaction
+     * @return The final files
+     */
+    private static StoredTabletFile[] applyTransaction(StoredTabletFile[] files,
+        DatafileTransaction transaction) {
+      Set<StoredTabletFile> newFiles = new HashSet<>(Arrays.asList(files));
+      transaction.apply(newFiles);
+      return newFiles.toArray(new StoredTabletFile[0]);
+    }
+
+    Date getInitialDate() {
+      return Date.from(Instant.ofEpochMilli(initialTs));
+    }
+
+    int getNumTransactions() {
+      return tabletLog.size();
+    }
+
+    List<DatafileTransaction> getTransactions() {
+      return tabletLog.toList();
+    }
+
+    Set<StoredTabletFile> getExpectedFiles() {
+      return new HashSet<>(Arrays.asList(finalFiles));
+    }
+
+    boolean isEmpty() {
+      return tabletLog.isEmpty();
+    }
+
+    /**
+     * Return a human readable dump of the log
+     *
+     * @param extent The tablet extent
+     * @return A dump of the log
+     */
+    String dumpLog(KeyExtent extent, boolean clear) {
+      // first lets get a consistent set of files and transactions
+      long initialTs;
+      StoredTabletFile[] initialFiles;
+      List<DatafileTransaction> transactions;
+      StoredTabletFile[] finalFiles;
+      boolean consistent;
+      do {
+        long updateCount = this.updateCount;
+        initialTs = this.initialTs;
+        initialFiles = this.initialFiles;
+        finalFiles = this.finalFiles;
+        transactions = this.tabletLog.toList();
+        consistent =
+            (updateCount == this.updateCount && updateCount == this.tabletLog.getUpdateCount());
```

Comment:
I think there are race conditions between this method and the setCapacity method that would be tricky to mitigate.  Seems like a synchronizing the methods on transactionLog would make it easier to reason about.  I doubt it would be a performance concern as long as not I/O is done while the locks are held.

Reply:
I will try to reason this out a little more.  The setCapacity is only done if the configuration is changed, so perhaps synchronization there would be ok.

Reply:
There is no race condition.  setCapacity creating an entirely new Ring object.  The setCapacity method is only called when adding a transaction which is already synchronized within a Tablet lock.